app/controllers/sign_up/recovery_codes_controller.rb

```diff
@@ -27,18 +27,16 @@ def new_code
 
     def confirm_has_not_already_viewed_recovery_code
       return if user_session[:first_time_recovery_code_view].present?
-      redirect_to after_sign_in_path_for(current_user)
+      redirect_to after_sign_in_path_for
     end
 
     def next_step
-      if session[:saml_request_url]
+      if session[:sp]
         sign_up_completed_path
```

Comment:
I think you might be right. The problem is that this controller is being called in various different contexts. It might be better to create a separate controller for each context. For example, one context is signing up. In that case, the logic will hold because we would know that the controller is only called during the sign up process, and then the logic would only depend on how the user came to the site before they started signing up (directly or via SP).

Currently, we are also using this controller in the scenario where an LOA3 users resets their password, but [the tests](https://github.com/18F/identity-idp/blob/master/spec/features/visitors/password_recovery_spec.rb#L270-L317) assume the user is coming to the site directly. We don't have tests that start the session by making a SAML request. If we did, I'm pretty sure they would fail because the user would be redirected to `sign_up_completed_path` instead of `reactivate_profile_path`.

Temporarily, I suppose we could reverse the order of this conditional such that `if current_user.password_reset_profile.present?` is first.